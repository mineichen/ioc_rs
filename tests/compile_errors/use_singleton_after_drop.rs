use ioc_rs::{Singleton, ServiceCollection};

fn main() {
    let mut container = ServiceCollection::new();
    container.register_singleton::<(), _>(|_| 0);
    let provider = container.build().expect("Expected to have all dependencies");
    let nr_ref = provider.get::<Singleton::<i32>>().unwrap();
    drop(provider); 
    assert_eq!(
        2, 
        *nr_ref
    );
}