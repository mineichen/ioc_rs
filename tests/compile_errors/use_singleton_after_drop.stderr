error[E0505]: cannot move out of `provider` because it is borrowed
  --> $DIR/use_singleton_after_drop.rs:8:10
   |
7  |       let nr_ref = provider.get::<Singleton::<i32>>().unwrap();
   |                    -------- borrow of `provider` occurs here
8  |       drop(provider);
   |            ^^^^^^^^ move out of `provider` occurs here
9  | /     assert_eq!(
10 | |         2,
11 | |         *nr_ref
12 | |     );
   | |______- borrow later used here
